DVCodingStyle.md

```diff
@@ -0,0 +1,864 @@
+    if (!uvm_config_db#(...)::get(...)) begin
+      `uvm_fatal(...)
+    end
+    ```
+6.  Wildcards must not be used in the field argument of `uvm_config_db::set(...)` calls.
+    It is allowed to use wildcards in the inst_name argument of these calls,
+    with proper precautions.
+
+
+### Sequences
+
+1.  It is recommended to keep sequences as generic as possible and to avoid
+    writing directed sequences unless absolutely necessary.
+2.  It is recommended to avoid explicit delays in terms of absolute time to
+    make code emulation friendly. Delays in terms of clock cycles are allowed.
+3.  The `body()` method should only execute the raw functional behavior of
+    the sequence.
+    Any associated housekeeping code should be placed elsewhere, such as in the
+    `pre_start()` and `post_start()` methods.
+    *   `pre_start()` and `post_start()` must be used instead of `pre_body()`
+        and `post_body()`, as these will always be called, even for subsequences
+        called from a parent sequence.
+4.  When creating a sequence object, always randomize the sequence before starting it.
+5.  Virtual sequences must be used to coordinate the behavior of multiple agents and multiple sequences.
+6.  Virtual sequences must be started on the null sequencer or on a valid virtual sequencer.
+
+
+### Sequence Items
+
+It is recommended, but not necessary, to create small "unit tests" for
+transactions objects to ensure the randomization constraints are working as
+intended.
+
+
+### RAL Usage
+
+When using the UVM RAL model, care must be taken when using `uvm_reg::set()` and
+`uvm_reg::update()`, as these are both non-atomic operations that can easily
+cause race conditions if there are any parallel threads.
+
+
+### Objections and Coordinating End of Test
+
+1.  At a high level, a phase should not end when either there is more stimulus
+    to send during that phase or the DUT has yet to respond to some stimulus that
+    has been sent.
+    To prevent a phase from ending in the former case, raise and lower objections
+    inside tests.
+    To prevent a phase from ending in the latter case, raise objections in the
+    scoreboard whenever expect queues are not empty and lower them when they are empty.
+2.  It is recommended to avoid using objections in drivers or monitors
+    for performance reasons.
+3.  If a sequence needs to raise/drop objections, it should call
+    `raise_objection` in its `pre_start()` method and `drop_objection` in its
+    `post_start()` method. A good alternative to this is to call
+    `set_automatic_phase_objection(1)` in the sequence constructor.
+4.  Always perform the test `if (starting_phase != null)` before calling
+    `raise_objection` or `drop_objection` within a sequence. Prior to uvm-1.2,
+    `starting_phase` was a member of the class `uvm_sequence_base`.
+    From uvm-1.2 onward, the `starting_phase` variable is deprecated and instead
+    must be accessed using the `get_starting_phase` method.
```

Comment:
Do we have a version of UVM that we'll always target? If >= 1.2, this can be simplified.

Reply:
@sriyerg what are your thoughts about this phrasing?

Reply:
At the moment we are using UVM 1.2 with both simulators, so wouldn't this be flagged as a compile error (and hence force the writer to update the code anyway)? I don't think we will ever revert back to an older version.

Reply:
Updated to solely reflect usage of the `get_starting_phase()` method.